### ISSUE TITLE
2.0 SessionAuthenticator crashing

### ISSUE THREAD

**Issue by gsechan** [couchbase/couchbase-lite-android#1346]
Running the following code:

            URI uri = null;
            try {
                uri = new URI("blips://gabe-couch.dev.handshake.com:4984/db");
            } catch (URISyntaxException e) {
                e.printStackTrace();
            }
            ReplicatorConfiguration replConfig = new ReplicatorConfiguration(database, uri);
            replConfig.setAuthenticator(new SessionAuthenticator(sessionInfo.cookieValue, sessionInfo.cookieDate, sessionInfo.cookieName));
            replConfig.setContinuous(false);
            replConfig.setReplicatorType(ReplicatorConfiguration.ReplicatorType.PULL);
            replicator = new Replicator(replConfig);
            replicator.start();

I'm crashing after calling start with the following stack trace:

08-14 16:25:37.675 542-6129/com.handshake.hsdm E/AndroidRuntime: FATAL EXCEPTION: Thread-16
                                                                 Process: com.handshake.hsdm, PID: 542
                                                                 java.lang.ClassCastException: java.lang.String cannot be cast to com.couchbase.litecore.fleece.FLValue
                                                                     at com.couchbase.lite.internal.replicator.CBLWebSocket.newRequest(CBLWebSocket.java:326)
                                                                     at com.couchbase.lite.internal.replicator.CBLWebSocket.start(CBLWebSocket.java:171)
                                                                     at com.couchbase.litecore.C4Socket.open(C4Socket.java:121)

It looks like either the problem is you should expect a String there, or whatever code you have in JNI that writes the options map back to Java from the core needs to populate that as an FLValue.

This crash was seen in DP12 and DP13.

**Comment by AlBlanc** [comment]
Same here. 

The app did not crash on the first opening and authenticated well. But since, It crashes all the time.

**Comment by AlBlanc** [comment]
Any recommendation to avoid this bug?

It is a blocking issue for our project, we had to disable sync, we can't go further and DB16 seems so far. Any help would be really appreciated.

**Comment by gsechan** [comment]
@AlBlanc  If you still need a workaround, replace newRequest in CBLWebSocket with:  

    private Request newRequest() {
        Request.Builder builder = new Request.Builder();

        // Sets the URL target of this request.
        builder.url(uri.toString());

        // Set/update the "Host" header:
        String host = uri.getHost();
        if (uri.getPort() != -1)
            host = String.format(Locale.ENGLISH, "%s:%d", host, uri.getPort());
        builder.header("Host", host);

        // Configure the nonce/key for the request:
        byte[] nonceBytes = new byte[16];
        new Random().nextBytes(nonceBytes);
        String nonceKey = encodeToString(nonceBytes, NO_WRAP);

        // Construct the HTTP request:
        if (options != null) {
            FLValue value = (FLValue) options.get(kC4ReplicatorOptionExtraHeaders);
            if (value != null && value.asDict() != null) {
                for (Map.Entry<String, Object> entry : value.asDict().entrySet())
                    builder.header(entry.getKey(), entry.getValue().toString());
            }
            // Cookies
            Object objvalue = options.get(kC4ReplicatorOptionCookies);
            if(objvalue != null) {
                String stringValue = null;
                if (objvalue instanceof String) {
                    stringValue = (String) objvalue;
                } else if (objvalue instanceof FLValue) {
                    stringValue = ((FLValue) objvalue).asString();
                }
                if (stringValue != null)
                    builder.addHeader("Cookie", stringValue);
            }
        }

        // other header values
        builder.header("Connection", "Upgrade");
        builder.header("Upgrade", "websocket");
        builder.header("Sec-WebSocket-Version", "13");
        builder.header("Sec-WebSocket-Key", nonceKey);

        // Add User-Agent if necessary:
        builder.header("User-Agent", getUserAgent());

        return builder.build();
    }

Basically the same code, but it accepts the value could be a string.  I'm sure this isn't quite the right way to fix it permanently, but its how I'm moving forward.

**Comment by hideki** [comment]
@AlBlanc sorry for taking a long time to address this problem.
@gsechan Thank you for your solution.
